src/coreclr/jit/morph.cpp

```diff
@@ -10870,6 +10870,97 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
             return node;
         }
 
+#if defined(TARGET_XARCH) || defined(TARGET_ARM64)
+
+#if defined(TARGET_XARCH)
+        case NI_SSE_And:
+        case NI_SSE2_And:
+        case NI_AVX_And:
+        case NI_AVX2_And:
+#elif defined(TARGET_ARM64)
+        case NI_AdvSimd_And:
+#endif
+        {
+            assert(node->GetOperandCount() == 2);
+
+            GenTree*            op1      = node->Op(1);
+            GenTree*            op2      = node->Op(2);
+            GenTree*            lhs      = nullptr;
+            GenTree*            rhs      = nullptr;
+            GenTreeHWIntrinsic* inner_hw = nullptr;
+
+            // Transforms ~v1 & v2 to VectorXxx.AndNot(v1, v2)
+            if (op1->OperIs(GT_HWINTRINSIC))
+            {
+                GenTreeHWIntrinsic* xor_hw = op1->AsHWIntrinsic();
+                switch (xor_hw->GetHWIntrinsicId())
+                {
+#if defined(TARGET_XARCH) || defined(TARGET_ARM64)
+
+#if defined(TARGET_XARCH)
+                    case NI_SSE_Xor:
+                    case NI_SSE2_Xor:
+                    case NI_AVX_Xor:
+                    case NI_AVX2_Xor:
+#elif defined(TARGET_ARM64)
+                    case NI_AdvSimd_Xor:
+#endif
+                        inner_hw = xor_hw;
+                        rhs      = op2;
+#endif
+                    default:
+                    {
+                        break;
+                    }
+                }
+            }
+
+            // Transforms v2 & (~v1) to VectorXxx.AndNot(v2, v1)
+            if (op2->OperIs(GT_HWINTRINSIC))
```

Comment:
This check is going to miss the opt if we have something like `((x ^ AllBitsSet) & (y ^ z)`. Such a tree could have been transformed into `AndNot((y ^ z), x)`

In general you're going to need to match `(op1 ^ AllBitsSet)` up front before determining if its a match and then if that fails do the same check for `(op2 ^ AllBitsSet)`.

For Arm64, you'll also need to directly check for `~op1` or `~op2` (since `NI_AdvSimd_Not` exists).

There are some things we could do to make this overall simpler, but they are slightly more involved changes.

Reply:
I'd, in general, recommend extracting some of this to a helper.

For example, you could define something like:
```cpp
genTreeOps GenTreeHWIntrinsic::HWOperGet()
{
    switch (GetHWIntrinsicId())
    {
#if defined(TARGET_XARCH)
        case NI_SSE_And:
        case NI_SSE2_And:
        case NI_AVX_And:
        case NI_AVX2_And:
#elif defined(TARGET_ARM64)
        case NI_AdvSimd_And:
#endif
        {
            return GT_AND;
        }

#if defined(TARGET_ARM64)
        case NI_AdvSimd_Not:
        {
            return GT_NOT;
        }
#endif

#if defined(TARGET_XARCH)
        case NI_SSE_Xor:
        case NI_SSE2_Xor:
        case NI_AVX_Xor:
        case NI_AVX2_Xor:
#elif defined(TARGET_ARM64)
        case NI_AdvSimd_Xor:
#endif
        {
            return GT_XOR;
        }

        // TODO: Handle other cases

        default:
        {
            return GT_NONE;
        }
    }
}
```

Such a helper allows you to instead switch over the `genTreeOps` equivalent. So you could have something like:
```
switch (node->HWOperGet())
{
    case GT_AND:
    {
        GenTree* op1 = node->Op(1);
        GenTree* op2 = node->Op(2);
        GenTree* lhs = nullptr;
        GenTree* rhs = nullptr;

        if (op1->OperIsHWIntrinsic())
        {
            // Try handle: ~op1 & op2
            GenTreeHWIntrinsic* hw     = op1->AsHWIntrinsic();
            genTreeOps          hwOper = hw->HWOperGet();

            if (hwOper == GT_NOT)
            {
                lhs = op2;
                rhs = op1;
            }
            else if (op1Oper == GT_XOR)
            {
                GenTree* hwOp1 = hw->Op(1);
                GenTree* hwOp2 = hw->Op(2);

                if (hwOp1->IsVectorAllBitsSet())
                {
                    lhs = op2;
                    rhs = hwOp2;
                }
                else if (hwOp2->IsVectorAllBitsSet())
                {
                    lhs = op2;
                    rhs = hwOp1;
                }
            }
        }

        if ((lhs == nullptr) && op2->OperIsHWIntrinsic())
        {
            // Try handle: op1 & ~op2
            GenTreeHWIntrinsic* hw     = op2->AsHWIntrinsic();
            genTreeOps          hwOper = hw->HWOperGet();

            if (hwOper == GT_NOT)
            {
                lhs = op1;
                rhs = op2;
            }
            else if (op1Oper == GT_XOR)
            {
                GenTree* hwOp1 = hw->Op(1);
                GenTree* hwOp2 = hw->Op(2);

                if (hwOp1->IsVectorAllBitsSet())
                {
                    lhs = op1;
                    rhs = hwOp2;
                }
                else if (hwOp2->IsVectorAllBitsSet())
                {
                    lhs = op1;
                    rhs = hwOp1;
                }
            }
        }

        if (lhs == nullptr)
        {
            break;
        }

        GenTree* andnNode = gtNewSimdBinOpNode(GT_AND_NOT, simdType, lhs, rhs, simdBaseJitType, simdSize, true);

        DEBUG_DESTROY_NODE(node);
        INDEBUG(andnNode->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);

        return andnNode;
    }

    default:
    {
        break;
    }
}
```

You could of course also extract the `NOT op` vs `op XOR AllBitsSet` matching logic to reduce duplication as well.

Reply:
Longer term, I think we may want to introduce a "fake" `Isa_Not` hwintrinsic id for xarch. That would allow morph to transform `x ^ AllBitsSet` into `Isa_Not` and then would in turn allow this case to be simplified in its pattern checks.

We may also want to normalize cases like `Sse_Xor`, `Sse2_Xor`, and `AdvSimd_Xor` into `Vector128_Xor`, so we don't need to consider xplat differences. But that will also involve significant refactorings, far more so than introducing a `HWOperGet` helper for the time being.